packages/apollo-server-plugin-base/src/index.ts

```diff
@@ -32,6 +32,17 @@ export interface GraphQLRequestListener<TContext = Record<string, any>> {
       'document' | 'operationName' | 'operation'
     >,
   ): ValueOrPromise<void>;
+  // If this hook is defined, it is invoked immediately before GraphQL execution
+  // would take place. If its return value resolves to a non-null
+  // GraphQLResponse, that result is used instead of executing the query.
+  // Hooks from different plugins are invoked in series and the first non-null
+  // response is used.
+  responseForOperation?(
+    requestContext: WithRequired<
+      GraphQLRequestContext<TContext>,
+      'document' | 'operationName' | 'operation'
```

Comment:
In the same spirit as one of my previous comments, I think `metrics` is guaranteed to be here at this point, right?  Could we declare `metrics` as required and avoid some not-null assertions? 

```suggestion
      'document' | 'operationName' | 'operation' | 'metrics'
```

Reply:
Honestly I'd like to declare it as required overall ... it's kinda a shame that GraphQLRequestContext has to represent both the argument to processGraphQLRequest *and* the argument to hooks it invokes. Maybe I could define some HookGraphQLRequestContext that's a WithRequired which at least has source and metrics set on it? Though maybe that's not great.

Reply:
We do have some progress we can make on improving this, but I'd not ask you to address it here.  There are other hooks that could benefit from something in this regard and @martijnwalraven 	and I have been musing about an `apollo-server-api` that could take the place of this entirely.

I do think some types could at least be shared between here and Apollo Server, but we want to be careful to allow plugins to have more flexibility and not be explicitly pinned to a specific Apollo Server version, thus the loose coupling and duplication.  Apollo Server could probably depend on this same type when it invokes the plugin, but that also has possibly similar undesirable couplings.  WIP, to say the least.